### DISCUSSION TITLE
How to trigger widget rerender?

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#6031] [Q&A] [answered]
I have a simple widget as follow:

```ts
/**
title: $:/plugins/linonetwo/zx-script/zx-script-output.js
type: application/javascript
module-type: widget
*/

(function () {
  const Widget = require('$:/core/modules/widgets/widget.js').widget;

  class ZXOutputWidget extends Widget {
    /**
     * Lifecycle method: call this.initialise and super
     */
    constructor(parseTreeNode, options) {
      super(parseTreeNode, options);
      this.initialise(parseTreeNode, options);
    }

    get stateTiddlerTitle() {
      const title = this.getAttribute('title', 'aaa.mjs');
      const stateTiddlerTitle = `$:/state/linonetwo/zx-script/output/${title}`;
      return stateTiddlerTitle;
    }

    /**
     * Lifecycle method: Render this widget into the DOM
     */
    render(parent, nextSibling) {
      this.parentDomNode = parent;
      this.computeAttributes();

      // DEBUG: console
      console.log(`rerender`);

      const outputElement = this.document.createElement('article');
      outputElement.innerHTML = $tw.wiki.getTiddlerText(this.stateTiddlerTitle);
      parent.insertBefore(outputElement, nextSibling);
      this.domNodes.push(outputElement);
    }

    refresh(changedTiddlers) {
      return this.stateTiddlerTitle in changedTiddlers && changedTiddlers[this.stateTiddlerTitle]?.modified;
    }
  }

  exports['zx-script-output'] = ZXOutputWidget;
})();
```

I can see `refresh` gets called, and even it returns true, there is no rerender.

![截屏2021-09-16 上午12 28 32](https://user-images.githubusercontent.com/3746270/133472575-46d14730-03c6-4708-82dc-905ee4c4126f.png)

I wonder why, since transclude widget is just doing like this

https://github.com/Jermolene/TiddlyWiki5/blob/f72512369014499fd9ab4e550c6e7247669f3be6/core/modules/widgets/transclude.js#L111-L119

background:

I'm displaying [zx](https://github.com/google/zx)'s output in this widget using

```jsx
creator: LinOnetwo
tags: $:/tags/ViewTemplate
title: $:/plugins/linonetwo/zx-script/OutputBar
type: text/vnd.tiddlywiki

<$zx-script-output title={{!!title}} />
```

**Comment by linonetwo** [comment]
Unable to solve this... Use wikitext suggest by a friend:

```jsx
<$vars currentTiddler={{{[{!!title}addprefix[$:/state/linonetwo/zx-script/output/]] }}}>

{{||$:/core/ui/ViewTemplate/body}}

</$vars>
```

The output is pretty good, can even render the output as rendered tiddler.

<img width="801" alt="D77E6CAACDBA9773AB396A12602C9EE5" src="https://user-images.githubusercontent.com/3746270/133476496-3dc8ef0c-f3a4-4c08-b207-737eb62fb33f.png">

**Comment by saqimtiaz** [comment, marked as answer]
You need to call `refreshSelf()` inside `refresh()` when the conditions are met for needing a refresh. Just returning true does not trigger a refresh of the widget.

**Reply by saqimtiaz** [reply]
refresh(changedTiddlers) {
      if(this.stateTiddlerTitle in changedTiddlers && changedTiddlers[this.stateTiddlerTitle]?.modified) {
          this.refreshSelf();
          return true;
      }
    }

**Reply by linonetwo** [reply]
Thanks, I thought tw will do it for me :P

**Reply by saqimtiaz** [reply]
@linonetwo keep in mind that if your widget will have child widgets (that is other widgets inside it) then you also need to handle the situation where the widget does not need a refresh but children might:

```
refresh(changedTiddlers) {
  if(this.stateTiddlerTitle in changedTiddlers && changedTiddlers[this.stateTiddlerTitle]?.modified) {
      this.refreshSelf();
      return true;
  } else {
      return this.refreshChildren(changedTiddlers);
  }
}
```

**Reply by linonetwo** [reply]
Thank you, this is helpful, since I've just posted another question that I want to render a child widget in my widget.

**Reply by linonetwo** [reply]
do you mean `return this.refreshChildren(changedTiddlers);`?

**Reply by saqimtiaz** [reply]
Yes, correct.
